### PR TITLE
Number field revamp

### PR DESCRIPTION
**Pull request by theofficialnar**
[number-field.webm](https://github.com/zesty-io/manager-ui/assets/28705606/a578d0bd-c05a-418c-9608-b5912b86c0dc)

### CODE REVIEW

File: package.json
Comment (theofficialnar) [reply package.json:90]: Since the package already handles the input pattern matching, value conversion (which basically strips out the , separators) and formatting all in one without having to do all those manually on each input change.

File: src/shell/components/FieldTypeNumber.tsx
Comment (theofficialnar) [reply src/shell/components/FieldTypeNumber.tsx]: I've updated how the increment/decrement works to make sure decimals are saved properly without making use of toFixed

**Comment by zcolah** [comment]
@agalin920 

**> Primarily how does a user know whether the number is being saved with or without the commas?**
Good question. We will add an information note to the number field docs to mention that they do not get saved with commas. We have multiple fields where what the marketer sees is not what the developer sees and this will be similar to that. e.g. The WYSIWYG field, the Boolean field, etc.

**> 1. Not all regions make use of comma separators in the same way. (e.g periods are used to separate thousands and commas for decimals)**
This is a fair point and was considered. Since we are a U.S. based company with a majority of our clients also U.S. based, we decided to follow the US Number format. That being said in the future, if the need arises we will add the functionality for a user to define their number format. We have decided that this is out of scope for this project. 

> 2. Screen reader cannot read this number properly
Could you provide me with more context as to why you believe it cannot read the number properly? What is preventing it from happening? Any Loom video or supporting evidence for it would be appreciated.

**Comment by zcolah** [comment]
**@agalin920 I also wanted to give some key context on why the number separators (commas) are being introduced.** 

Currently when a user inputs a larger number (e.g. 1000000000) they tend to have difficulty reading and understanding what the value is. This leads to a user potentially misreading the number or adding an extra zero or a lesser zero by mistake. 
This issue can be very concerning for an insurance company or any large enterprise company where data integrity is very critical. 

https://github.com/zesty-io/manager-ui/issues/2456

**Comment by zcolah** [comment]
@agalin920 @theofficialnar I also do not have full context on why the entire solution being used to achieve the commas is bad but here are some discussions I have found online that discuss how others have achieved this. Maybe they could be helpful. Let me know what I can do to unblock this issue moving forward. 

1. https://community.wappler.io/t/is-there-an-easy-way-to-auto-format-input-field-with-thousand-separator-comma/39138/5
2. https://stackoverflow.com/questions/31867551/html-input-type-number-thousand-separator
3. https://github.com/amiryxe/easy-number-separator

**Comment by zcolah** [comment]
@zcolah while I understand the pain points you present I fundamentally disagree with the solution and believe the trade offs of ambiguous comma separation is worse. I agree with the html numeric input standard to not comma separate and even data entry software like sheets or excel also does not by default auto comma numbers. It makes data entry standard and in my opinion comma separation should be used for display purposes only and not data entry.

@agalin920 
> **data entry software like sheets or excel also does not by default auto comma numbers**
For all cells that are number types, Excel adds commas and decimal places actually by default. It does not add commas for plain text (string type) cells. In fact in most calculator interfaces as well commas are added by default in the number input. 

> **should be used for display purposes only**
The Content App, displays to a marketer, critical number values in their content items during content creation and content review/editing. 

During content creation, it is important for us to display the number they are inputting in the most readable easy to scan format so as to ensure they do not make a mistake. For example it can be a question of them inputting 100,000 vs 1,000,000 for a critical page. 

During content review and editing, we need to make it as easy as possible for them to read a number in a number input. 
Compare reading and scanning 1,000,000,000 vs 1000000000. A user can be more likely to make a mistake reading the number without commas.

**Comment by zcolah** [comment]
@agalin920 @theofficialnar 
I do want to help. I found these two Codepen examples. Potentially we could see how they did it and maybe then not use an external package/dependency? 

https://codepen.io/vladracoare/pen/VwLRqRy
https://codepen.io/tsunet111/pen/GbpwZa

**Comment by zcolah** [comment]
@agalin920 

> Not really looking for solution examples, even without a direct dependency they all make use of JS to manipulate the input. The main flag im raising is html number input standard deviation.

As a content management system, we have and we will sometimes deviate from the default inputs. We have done this for fields in the past as well. This will be not just in terms of enhancing their visual appearance but also their functionality. This is not new to the industry and nor are we pioneers. We are just improving existing the number field to solve for a critical user problem described above. 

I am ready to look for solutions to solve the user problem. If you are not, then we need to accept @theofficialnar's solution and move forward and close the ticket and accept it as an experiment and product decision.

**Comment by zcolah** [comment]
@agalin920 

It is very important we put ourselves in the shoes of someone inputting the weight of an item or a shipment or a policy etc. One single digit mistake can lead to a lot of losses for an enterprise company such as an e-commerce company or insurance company. It hurts those who are at scale.

**Comment by zcolah** [comment]
@agalin920 

In the future we would give the user the option to use European number notations or switch off the comma separators via the rules tab in Schema app. We are not saying that this will always be the only way numbers will be presented. We are solving for a critical user concern. That is all.

**Comment by zcolah** [comment]
@theofficialnar 

Tiny VQA issue to resolve: https://www.figma.com/file/zOI7oSH3fG1XDmmPzsP6PX/Zesty-3.0-Revamp?type=design&node-id=4015%3A241611&mode=dev&t=WrGlDxwAh1Efh4z0-1

File: src/shell/components/FieldTypeNumber.tsx
Comment (theofficialnar) [reply src/shell/components/FieldTypeNumber.tsx:60]: This actually resolves a different issue other than the default to 0 when empty. The cursor jumps before 0 instead of being after (which is how the input type number behaves) if I do that that's why I ended up just preventing the backspace and delete keys when the value is already 0.

**Comment by theofficialnar** [comment]
> @theofficialnar
> 
> * [ ]  Tiny VQA issue to resolve: https://www.figma.com/file/zOI7oSH3fG1XDmmPzsP6PX/Zesty-3.0-Revamp?type=design&node-id=4015%3A241611&mode=dev&t=WrGlDxwAh1Efh4z0-1

It already has an 8px gap

**Comment by theofficialnar** [comment]
> There is a bug when inputing very large numbers and hitting save. It gets set as a completely different number and in decimal.
> 
> looks like increment and decrement also causing this
> 
> @zcolah @theofficialnar

@agalin920 so apparently big numbers are getting converted after they're saved, for example I updated the field with this value _1231233453453453456756756756756234089089089089089089089089089056456456454566666666777788678890890890_ and once saved and rtk query refreshes the data, the api responds with a exponential notation number for big numbers which is _1.2312334534534534e+99_ but then the **new** number field removes the "e+99" that's why you're seeing the error.

What I could do is treat all number inputs as numbers to begin with but once you start typing in more than 15-16 digits, javascript rounds it up causing all other inputs to be trailing 0s. Like in the example I mentioned above, the api's response which is _1.2312334534534534e+99_ when converted via `toLocaleString` becomes _1,231,233,453,453,453,400,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000_ due to it exceeding Number.MAX_SAFE_INTEGER.

I guess, if we **really** want to save the number the user inputs exactly, then we'd probably have to make sure that the api saves it as a string, but then again it sort of defeats the purpose of a number field since you'd expect it to be storing numbers.

**Comment by zcolah** [comment]
Potentially an alternate solution is we could give an error to a user that
prevents them from inserting such a large number. Similar to how we have a
max character count for single line text field, we could have a max number
value. Just an idea.

**Comment by zcolah** [comment]
Okay that is fair. 

1. I believe we should not accept any number that goes beyond Number.MAX_SAFE_INTEGER. We should throw an error for any number that exceeds that. 

2. Idea: We move forward with Nar's solution of rounding numbers and provide a warning to a user explaining to them that the end of the number is being rounded and write logic to correctly interpret the number in exponent notation (1.2312334534534534e+99) the database API returns to us. This is not most ideal though. Else we save it as a string and show it exactly as it is. 

Also @agalin920 you are correct that there can be users who have saved very large numbers in our system. But we cannot say that for sure. Potentially we could ask @markelmad to let us know what is the largest number saved in the system till date.

**Comment by theofficialnar** [comment]
@agalin920 @zcolah I'm currently trying to work on something that would make the UI show the actual exponents instead of the trailing 0s, just fixing some minor issues.

File: src/shell/components/NumberFormatInput/index.tsx
Comment (theofficialnar) [reply src/shell/components/NumberFormatInput/index.tsx:12]: I purposely left this for now until we get the final approval from @zcolah's vqa. I'll remove this file and the dependency once the new solution is approved

**Comment by zcolah** [comment]
@agalin920 @theofficialnar May you review this video that shares some insights from my testing of the number field? 
https://www.loom.com/share/3f0e4faa132748ffa78f5a7d8a528119?from_recorder=1&focus_title=1

**Comment by giseleblair** [comment]
I strongly agree with @agalin920 that we should not be saving commas into the values. This field is intended save integer values. Which it is currently not doing. It is saving strings. We need to better scope how the values are to be used by the customer. The number value is intended to be used programmatically for sorting or filtering which is why it is important to maintain the values as integers. This would remove allowing any commas or negative symbols. If we want to save float integers, that should be an entirely separate field type. The user experience needs to be better defined as to what kind of value we are saving and then have the interface reflect that.

**Comment by giseleblair** [comment]
For further documentation please read our API spec on the "number" field datatype: https://instances-api.zesty.org/#689c935d-a46c-8eef-cf20-df60f55c38d3

The API allows for an integer or a float value. However, the interface never defaults to save an integer value, it always saves a string. Adding commas to the values should be out of the question for the saved value. The interface can display it but it should never save it via the API. If the user wants to save a negative or exponent value, then a float type will be necessary and it will need to be saved as a string. Otherwise, the default should always be an integer. 
Saving integers vs float values can eventually be one of the "rules" set on the field. But I know we are not quite there yet. But setting the behavior now for that future use case is important.

**Comment by zcolah** [comment]
@giseleblair 

1. We are saving the value as a float to the database. 
2. This float value contains no commas. This is only in the presentation layer to improve the readability of the number. 
3. The number field does support negative values as well. 
4. The reason why you see it represented as a string in the JSON is more likely an issue with Web Engine. @markelmad can you confirm that the number field is saved as a number? Reference links: 

- https://8xbq19z1-dev.preview.stage.zesty.io/-/instant/7-deffefd282-gqqbm3.json
- https://8-f48cf3a682-7fthvk.manager.stage.zesty.io/content/6-556370-8sh47g/7-deffefd282-gqqbm3

**Comment by zcolah** [comment]
**@giseleblair also please note:** 
This number can be sorted and mathematical operations can be run on it. Whether the JSON returns it as a number is a Web Engine issue.

**Comment by giseleblair** [comment]
@zcolah the JSON returning it as a number is not a Webengine issue. The manager UI is explicitly making API requests that save these values as a string.

**Comment by theofficialnar** [comment]
@giseleblair where are you seeing it being saved as a string? I just checked just now and it's always saving it as a number.
image

**Comment by theofficialnar** [comment]
> @agalin920 @theofficialnar May you review this video that shares some insights from my testing of the number field? https://www.loom.com/share/3f0e4faa132748ffa78f5a7d8a528119?from_recorder=1&focus_title=1

@zcolah for the first concern where you can't update the exponent value and use the +/- buttons I just mimicked how the current number field in production behaves where it clears out the input field when a user changes an exponent number. Regarding the +/- buttons this is because adding and subtracting 1 no longer does anything in exponent numbers because like what I've pointed out on the previous comments, exponent numbers are always getting rounded up to 0s so adding and subtracting 1 won't really do anything anymore.

For the second issue, that's a valid point and I'll look into it.

**Comment by zcolah** [comment]
@theofficialnar Thank you for that clarification.

**Review by zcolah — CHANGES_REQUESTED** [review]
@theofficialnar caught a new bug related to decimal places in VQA which did not exist before: https://www.loom.com/share/b14736392b2940ec83fc86c158895510

Sorry for the delay on the VQA for this as I was in the hospital

**Review by zcolah — COMMENTED** [review]
Hi @theofficialnar 

- [x] A new bug has crept in since I last checked the number field. This time I cannot remove a decimal place if there is only a 0 right after it. Please see Loom for more context:

https://www.loom.com/share/b1d1d0f45bce45c3bf53757a69b01b49

**Comment by zcolah** [comment]
> * [ ]  @theofficialnar caught a new bug related to decimal places in VQA which did not exist before: https://www.loom.com/share/b14736392b2940ec83fc86c158895510

did you see this btw @theofficialnar

**Comment by theofficialnar** [comment]
> > * [ ]   @theofficialnar caught a new bug related to decimal places in VQA which did not exist before: https://www.loom.com/share/b14736392b2940ec83fc86c158895510
> 
> did you see this btw @theofficialnar

Yeah, this should already be fixed with the changes yesterday

**Review by zcolah — CHANGES_REQUESTED** [review]
@theofficialnar 

- [ ] I cannot remove 0 if it is the last digit in the decimal places: https://www.loom.com/share/53e547912e604b93a2a12123531101e2

- [ ] I cannot remove 0 if it is the last digit in the whole number places: https://www.loom.com/share/87cb53a9df684d37ad8533cf7ac02f5f

**Comment by zcolah** [comment]
@jomarmontuya 

- [ ] 1. Reduce the lag in showing the current date when the user opens the modal for the first time
- [ ] 2. If possible please add a cursor indicator so that the user knows they are typing and knows where they are typing
- [ ] 3. Add ability to update the entire date when it is highlighted/selected
- [ ] 4. Once a user has selected a date via the Calendar pop up, the date input can go back to it's default non selected state. It does not need to be in active state and the year does not need to be highlighted. 
- [ ] 5. If there is an invalid date, then please immediately display an error message to a user. [See Figma design.](https://www.figma.com/file/zOI7oSH3fG1XDmmPzsP6PX/Zesty-3.0-Revamp?type=design&node-id=4080%3A27763&mode=dev&t=sOKSk2D0KnWplKAw-1)

**Review by zcolah — CHANGES_REQUESTED** [review]
Wrong comment. Ignore.

**Comment by giseleblair** [comment]
> @giseleblair where are you seeing it being saved as a string? I checked just now and it's always saving it as a number. 

I am seeing that the number is being returned as an integer in the API. Looks like webengine is converting these to string in our read-only APIs. This is something we need to confirm with @jjaguilar08

File: src/shell/components/FieldTypeNumber.tsx
Comment (theofficialnar) [reply src/shell/components/FieldTypeNumber.tsx:48]: I’m actually no longer using the react-number format library here hence all the complexity and bugs since I have to handle all the cursor position etc manually. I’m actually considering using the library again since it takes care of all the formatting and resolves all of the bugs @zcolah has been raising for the past couple of days. Only downside to that library being that it automatically converts the exponents to numbers with leading 0s instead of showing the actual exponent.

**Comment by theofficialnar** [comment]
@zcolah @agalin920 what are your thoughts on bringing back in the library which handles the number formatting etc? That library resolves all the issues that are being brought up and the complexity of this component, only issue is that it automatically converts large exponent numbers to numbers with trailing 0s instead of showing the actual exponent.

**Comment by zcolah** [comment]
@theofficialnar fine by me as well

**Comment by zcolah** [comment]
@theofficialnar the Decimal places get rounded when increasing the number with the plus button: https://www.loom.com/share/0f4218027df54907a55d53e73ed0ddcb?from_recorder=1&focus_title=1

**Comment by theofficialnar** [comment]
@zcolah turns out it's a limitation with Javascript's floating-point precision hence it's getting rounded off, javascript numbers can hold 15 digits accurately

**Comment by zcolah** [comment]
@theofficialnar Thank you for that insight. It is fair to not resolve it but have a few follow up questions regarding the scope of this javascript constraint. 

1. Does this impact only a situation where a user tries to increase the value with the plus and minus buttons? 
2. Does this mean any value that has more than 15 digits will get rounded post that and saved like that in the database? 
3. What happens in a situation where the database returns a value with more than 15 digits? Does the input round it.

**Comment by theofficialnar** [comment]
@zcolah 
> Does this impact only a situation where a user tries to increase the value with the plus and minus buttons?
As of the moment, user input gets retained but the value saved on the database is rounded up.

> Does this mean any value that has more than 15 digits will get rounded post that and saved like that in the database?
Yes, unless they start exceeding the MAX_SAFE_INTEGER then they'll be converted to exponents

> What happens in a situation where the database returns a value with more than 15 digits? Does the input round it.
Yes

**Comment by zcolah** [comment]
> @zcolah
> 
> > Does this impact only a situation where a user tries to increase the value with the plus and minus buttons?
> 
> As of the moment, user input gets retained but the value saved on the database is rounded up.
> 
> > Does this mean any value that has more than 15 digits will get rounded post that and saved like that in the database?
> 
> Yes, unless they start exceeding the MAX_SAFE_INTEGER then they'll be converted to exponents
> 
> > What happens in a situation where the database returns a value with more than 15 digits? Does the input round it.
> 
> Yes

Thank you for these insights. I am going to give it a VQA label now.

**Pull request merged**
